### DISCUSSION TITLE
Duplicate commit created of not changed entity?

### DISCUSSION THREAD

**Discussion by edbras** [javers/javers#1330] [General]
Please provide some thoughts and inspiration for the following weird difference.

I am using the latest Javers version (7.x) and I am auditing dto's with an @Id. I have created a custom comparator for the OffsetDatetime to truncate Ms and compare them.

I am using Spring boot 3.x with Hibernate 6.x. I perform manual commits  (javers.commit(..) instead of the annotations. 
In a @SpringBootTest that runs in a transaction through the @Transaction, I perform the following actions: 
create an EquipmentDto, change it, update it, not-change it, and then update it again.
The last action is to check that only a commit is created by Javers when something changes. I flush and clear the entity manager between the crud actions in the test.

The EquipmentDto is mapped from the Equipment Entity that has OffsetDatetime createdAt and modifiedAt fields that is automatically updated by Spring (@EnableJpaAuditing), through "@EntityListeners(AuditingEntityListener.class)"

When I run the test locally, no extra commit is created (I have 2: one initial and one for the update), however, when I run it in the Azure devops pipeline, it fails as there are 3 commits created.  Devops runs the same docker db version (Postgres 15) as locally and when I log the Diff javers.compare(update1, update2) it shows there are no differences (so the modifiedAt are the same). I also put a sleep of 1 sec before the last update and it keeps failing.
Below the 3 snapshots that appear in the logging. The last one is the same, apart from the modifiedAt OffsetDatetime, for some reason the precision is different: 
```
commit 3: 2023-10-15T10:53:49.020299Z
commit 2: 2023-10-15T10:53:49.020299085Z

```
I am not sure why it's difference, but could this cause the extra commit? I though not by using the Custom comparator. I log all the comparator actions and it seems to work well.

I have difficulty understanding why an extra commit is added and how to trace down the cause, please some help? 
I can copy/paste some code snippets if needed? 

```
Snapshot{commit:3.01, id:...EquipmentRestDto/bedbb554-c4fa-43ca-9e95-47620d23aa39, version:3, state:{createdAt:2023-10-15T10:53:48.954938Z, createdBy:test user, description:description_98ef4eca-19a6-458f-bc07-77b118031ac4, id:bedbb554-c4fa-43ca-9e95-47620d23aa39, modifiedAt:2023-10-15T10:53:49.020299Z, modifiedBy:test user, name:name_990d312b-2678-4f9f-bb3d-792c5e46cf8f, template:a396d8e7-30bb-45f7-8d09-c25f40651ad4, type:EQUIPMENT, version:1}}, 

Snapshot{commit:2.03, id:...EquipmentRestDto/bedbb554-c4fa-43ca-9e95-47620d23aa39, version:2, state:{createdAt:2023-10-15T10:53:48.954938Z, createdBy:test user, description:description_98ef4eca-19a6-458f-bc07-77b118031ac4, id:bedbb554-c4fa-43ca-9e95-47620d23aa39, modifiedAt:2023-10-15T10:53:49.020299085Z, modifiedBy:test user, name:name_990d312b-2678-4f9f-bb3d-792c5e46cf8f, template:a396d8e7-30bb-45f7-8d09-c25f40651ad4, type:EQUIPMENT, version:1}}, 

Snapshot{commit:1.05, id:...EquipmentRestDto/bedbb554-c4fa-43ca-9e95-47620d23aa39, version:1, state:{createdAt:2023-10-15T10:53:48.954938077Z, createdBy:test user, description:description_44a25f1a-314c-44e3-83b8-1ad96c3950b9, id:bedbb554-c4fa-43ca-9e95-47620d23aa39, modifiedAt:2023-10-15T10:53:48.954938077Z, modifiedBy:test user, name:Equipmenta1923ee1-c49e-4cc7-994a-c4a5490cce31, template:a396d8e7-30bb-45f7-8d09-c25f40651ad4, type:EQUIPMENT, version:0}}]]

```

**Comment by bartoszwalacik** [comment]
I'm afraid you need to do the debugging on your own. Are you sure your custom comparator is used? Are you sure you have the same time zone settings locally and on Azure, are you sure that your local db configuration is exactly the same like on Azure, can you reproduce the issue locally?

**Reply by edbras** [reply]
Thanks for the input. 
I found it, I encountered 2 issues: 
- Different time precision on different platforms (depends on the os clock): on mijn OSX I have no Nanon seconds (it's rounded to microsecond) and on the linux docker we use, we have nanoseconds. So locally the test did not fail as the difference in the nano part.
 
- My comparator contained a little bug: simple truncating of the nano part was not good enough, you need to round it correctly, else these times are not the same: 
2023-10-15T10:53:49.020299Z
2023-10-15T10:53:49.020298805Z
(I think the nano part is different due to retrieval from the db)

Anyway, it works now, it cost some time, but I found it, thanks for the tips.
